### PR TITLE
Questions by ticket types and badge feature types

### PR DESCRIPTION
**Pull request by romanetar**
ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3241447&fileview=grid

Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/Models/Foundation/Summit/Registration/ExtraQuestions/SummitOrderExtraQuestionType.php [hunk 1]
```diff
@@ -102,12 +125,47 @@ public function getExternalId(): ?string
...
+        if ($this->allowed_ticket_types->contains($ticket_type)) return;
+        $this->allowed_ticket_types->add($ticket_type);
+    }
+
+    /**
+     * @return SummitBadgeFeatureType[]
+     */
+    public function getAllowedBadgeFeatureTypes(): array
+    {
+        return $this->allowed_badge_features_types->toArray();
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Registration/ExtraQuestions/SummitOrderExtraQuestionType.php:157]: this is not very performant way bc its allocates the entire collection and u lost the lazy load collection feature

File: app/Models/Foundation/Summit/Registration/ExtraQuestions/SummitOrderExtraQuestionType.php [hunk 2]
```diff
@@ -102,12 +125,47 @@ public function getExternalId(): ?string
...
     {
         $this->external_id = $external_id;
     }
 
+    /**
+     * @return SummitTicketType[]
+     */
+    public function getAllowedTicketTypes(): array
+    {
+        return $this->allowed_ticket_types->toArray();
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Registration/ExtraQuestions/SummitOrderExtraQuestionType.php:140]: this is not very performant way bc its allocates the entire collection and u lost the lazy load collection feature

File: app/Models/Foundation/Summit/Registration/ExtraQuestions/SummitOrderExtraQuestionType.php [hunk 3]
```diff
@@ -51,6 +52,26 @@ class SummitOrderExtraQuestionType extends ExtraQuestionType
...
+     * @ORM\JoinTable(name="SummitOrderExtraQuestionType_SummitTicketType",
+     *      joinColumns={@ORM\JoinColumn(name="SummitOrderExtraQuestionTypeID", referencedColumnName="ID")},
+     *      inverseJoinColumns={@ORM\JoinColumn(name="SummitTicketTypeID", referencedColumnName="ID")}
+     *      )
+     * @var SummitTicketType[]
+     */
+    private $allowed_ticket_types;
+
+    /**
+     * @ORM\ManyToMany(targetEntity="SummitBadgeFeatureType", inversedBy="extra_question_types")
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Registration/ExtraQuestions/SummitOrderExtraQuestionType.php:66]: check https://www.doctrine-project.org/projects/doctrine-orm/en/2.14/tutorials/extra-lazy-associations.html

Comment (romanetar) [reply app/Models/Foundation/Summit/Registration/ExtraQuestions/SummitOrderExtraQuestionType.php]: done

File: app/Models/Foundation/Summit/Registration/ExtraQuestions/SummitOrderExtraQuestionType.php (hunk 2 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Registration/ExtraQuestions/SummitOrderExtraQuestionType.php]: done

File: app/Models/Foundation/Summit/Registration/ExtraQuestions/SummitOrderExtraQuestionType.php (hunk 1 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Registration/ExtraQuestions/SummitOrderExtraQuestionType.php]: done

File: app/ModelSerializers/ExtraQuestionTypeSerializer.php [hunk 4]
```diff
@@ -89,16 +97,19 @@ public function serialize($expand = null, array $fields = array(), array $relati
         'values' => [
             'type' => Many2OneExpandSerializer::class,
             'getter' => 'getValues',
-            'test_rule' => 'ModelSerializers\\ExtraQuestionTypeSerializer::testRule'
+            'test_rule' => 'ModelSerializers\\ExtraQuestionTypeSerializer::testRule',
+            'should_skip_rule' => 'ModelSerializers\\ExtraQuestionTypeSerializer::shouldSkip',
```
Comment (smarcet) [inline app/ModelSerializers/ExtraQuestionTypeSerializer.php:101]: here should_skip_rule is not needed

File: app/ModelSerializers/ExtraQuestionTypeSerializer.php [hunk 5]
```diff
@@ -43,6 +43,14 @@ public static function testRule($e){
         return $e->allowsValues();
     }
 
+    public static function shouldSkip($e, $params){
+        if (array_key_exists('attendee', $params)) {
+            $attendee = $params['attendee'];
+            return !$attendee->isAllowedQuestion($e);
```
Comment (smarcet) [inline app/ModelSerializers/ExtraQuestionTypeSerializer.php:49]: $e is a QuestionRule, u need to get the question instance

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitOrderExtraQuestionTypeApiController.php [hunk 6]
```diff
@@ -448,4 +470,66 @@ public function deleteSubQuestionRule($summit_id, $question_id, $rule_id)
...
         });
     }
+
+    /**
+     * @param $summit_id
+     * @return mixed
+     */
+    public function getOwnAttendeeAllowedExtraQuestions($summit_id)
+    {
+        try {
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitOrderExtraQuestionTypeApiController.php:480]: use ProcessRequest Trait

File: app/ModelSerializers/ExtraQuestionTypeSerializer.php (hunk 5 above)
Comment (romanetar) [reply app/ModelSerializers/ExtraQuestionTypeSerializer.php]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitOrderExtraQuestionTypeApiController.php (hunk 6 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitOrderExtraQuestionTypeApiController.php]: done

File: database/seeders/ApiEndpointsSeeder.php [hunk 7]
```diff
@@ -2880,7 +2898,25 @@ private function seedSummitEndpoints()
                 'http_method' => 'PUT',
                 'scopes' => [sprintf(SummitScopes::WriteSummitData, $current_realm)],
             ),
-           [
+            [
+                'name' => 'get-attendee-allowed-extra-questions',
```
Comment (smarcet) [inline database/seeders/ApiEndpointsSeeder.php:2902]: this endpoint need admin authz

File: routes/api_v1.php [hunk 8]
```diff
@@ -1438,6 +1445,12 @@
                         });
                     });
                 });
+
+                // allowed extra questions
+                Route::group(array('prefix' => 'allowed-extra-questions'), function () {
+                    Route::get('', 'OAuth2SummitOrderExtraQuestionTypeApiController@getAttendeeExtraQuestions');
```
Comment (smarcet) [inline routes/api_v1.php:1451]: this endpoint needs admin authz

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitOrderExtraQuestionTypeApiController.php [hunk 9]
```diff
@@ -444,4 +455,96 @@ public function deleteSubQuestionRule($summit_id, $question_id, $rule_id)
...
+                return [
+                    'id',
+                    'name',
+                    'label',
+                    'order',
+                ];
+            },
+            function ($filter) use ($summit, $attendee) {
+                if ($filter instanceof Filter) {
+                    $filter->addFilterCondition(FilterElement::makeEqual('summit_id', $summit->getId()));
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitOrderExtraQuestionTypeApiController.php:525]: lets review the filtering condition here
it should be something like
(has_badge_feature_types == false and has_ticket_types == false) or ( allowed_badge_feature_type_id in ( attendees features) OR allowed_ticket_type_id in ( attendees tickets types)

Comment (smarcet) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitOrderExtraQuestionTypeApiController.php:525]: also its missing these filters
class==MainQuestion and usage==Ticket

File: tests/OAuth2SummitOrderExtraQuestionTypeApiTest.php [hunk 10]
```diff
@@ -716,4 +767,282 @@ public function testGetAllMainBySummit(){
...
+            [],
+            [],
+            [],
+            $headers
+        );
+
+        $content = $response->getContent();
+        $questions = json_decode($content);
+        $this->assertTrue(!is_null($questions));
+        $this->assertEquals($questions->total, count($allowed_subquestion_labels) + 1); //allowed subquestions + allowed main question
```
Comment (smarcet) [inline tests/OAuth2SummitOrderExtraQuestionTypeApiTest.php:894]: this test result is not correct
pleas check response of https://github.com/fntechgit/fnEvent/blob/88414da1ef3960b41b2630bcca3bb12d112ab434/src/actions/summit-actions.js#L32
should only return the main questions ( roots ) so in this case is 2

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: database/seeders/ApiEndpointsSeeder.php (hunk 7 above)
Comment (romanetar) [reply database/seeders/ApiEndpointsSeeder.php:2902]: done

File: routes/api_v1.php (hunk 8 above)
Comment (romanetar) [reply routes/api_v1.php]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitOrderExtraQuestionTypeApiController.php (hunk 9 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitOrderExtraQuestionTypeApiController.php:523]: done

File: tests/OAuth2SummitOrderExtraQuestionTypeApiTest.php (hunk 10 above)
Comment (romanetar) [reply tests/OAuth2SummitOrderExtraQuestionTypeApiTest.php]: done

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments on tix
regards

**Review by smarcet — APPROVED** [review]
LGTM

**Pull request merged**